sdk/core/azure-core/azure/core/pipeline/transport/_base.py

```diff
@@ -220,18 +220,32 @@ class HttpRequest(object):
     :param str url: At least complete scheme/host/path
     :param dict[str,str] headers: HTTP headers
     :param files: Files list.
-    :param data: Body to be sent.
+    :param data: Body to be sent. If you want to send a json body, you should use the
+     `json` kwarg instead. We will handle json serialization for your data.
     :type data: bytes or str.
+    :keyword json: A JSON serializable object. Serializes your inputted object. Use this
+     instead of data if you wish for us to handle json serialization of your object for you.
+    :raises: TypeError
     """
 
-    def __init__(self, method, url, headers=None, files=None, data=None):
-        # type: (str, str, Mapping[str, str], Any, Any) -> None
+    def __init__(self, method, url, headers=None, files=None, data=None, **kwargs):
+        # type: (str, str, Mapping[str, str], Any, Any, Any) -> None
         self.method = method
         self.url = url
         self.headers = _case_insensitive_dict(headers)
         self.files = files
-        self.data = data
         self.multipart_mixed_info = None  # type: Optional[Tuple]
+        json_kwarg = kwargs.pop("json", None)
+        if data and json_kwarg:
```

Comment:
Let's say a ValueError

Reply:
What is `json_kwarg` is `{}`? Or False? Or data is 0? If we want to check (and raise) if the user specified these values, we can either check if the values are `in kwargs` or use a different sentinel value than None/compare fals:y values. 

Interestingly enough, I'm not sure if requests raises an exception if you provide both `data` and `json`....

Reply:
(and technically, `None`/`null` is valid json - so there is a difference between not providing the json keyword argument and providing it with a `None` value, so we should distinguish if any value (including `None) was provided or not regardless of if we want to throw or not)

Reply:
@johanste yeah, just checked `requests` behavior, and it allow us to pass in both, but it actually seems to ignore `json` when both are passed

![image](https://user-images.githubusercontent.com/43154838/107665269-49264280-6c5b-11eb-9b31-7c1893c3b042.png)

i'm not sure how closely we're trying to model `HttpRequest` after `requests`, but I think it'd be fair to follow `requests` here

Reply:
decided offline with @johanste and @lmazuel that we will raise a `ValueError`